elements/px-catalog-router/px-catalog-router.html

```diff
@@ -281,22 +277,148 @@
+              var part = parts[i].substring(0, queryStart);
+              if (part.length) newParts.push(part);
+              if (query.length) foundQueries.push(query);
+            } else {
+              newParts.push(parts[i]);
             }
-          }, 10);
+          }
+
+          return resolve({ path: newParts.join('/'), query: foundQueries.join('&') });
+        }.bind(this));
+      },
+
+
+      /**
+       * When the raw URL path is changed, ensure that it has a beginning and a
+       * trailing slash to make `app-route` happy with its parsing patterns.
+       *
+       * @param {String} rawPath
+       * @return {String} - The modified `rawPath` string
+       */
+      _ensureHasSlashes: function(rawPath) {
+        // Ensure beginning slash
+        rawPath = (rawPath.substring(0,1) !== '/') ? '/' + rawPath : rawPath;
+        // Ensure trailing slash if does not end in `.html`
+        rawPath = (rawPath.slice(-1) !== '/' && rawPath.slice(-5) !== '.html') ? rawPath + '/' : rawPath;
+
+        return rawPath;
+      },
+
+      /**
+       * Accepts a raw query string and returns it as a parsed object.
+       *
+       * Function taken from `iron-query-params` element @v0.8.9. Original code
+       * copyright (c) 2015 The Polymer Project Authors and released under the
+       * BSD license. See the following page for full license information and
+       * the original source code:
+       *
+       * https://github.com/PolymerElements/iron-location/blob/master/iron-query-params.html
+       *
+       * @param {String} paramString
+       * @return {Object}
+       */
+      _parseQueryString: function(paramString) {
+        var params = {};
+        // Work around a bug in decodeURIComponent where + is not
+        // converted to spaces:
+        paramString = (paramString || '').replace(/\+/g, '%20');
+        var paramList = paramString.split('&');
+        for (var i = 0; i < paramList.length; i++) {
+          var param = paramList[i].split('=');
+          if (param[0]) {
+            params[decodeURIComponent(param[0])] =
+                decodeURIComponent(param[1] || '');
+          }
         }
+        return params;
+      },
+
+      /**
+       * Run a series of promises to clean up the raw path.
+       *
+       * 1. Places query params in right position
+       *
+       * Given a route like this: `/#/home?theme=dark`
+       * Will produce a path like this: `/?theme=dark#/home/`
+       *
+       * Given a route like this: `/#/css/thing?theme=dark`
+       * Will produce a path like this: `/?theme=dark#/css/thing`
+       *
+       * Given a route like this: `/#/css/thing/?theme=dark`
+       * Will produce a path like this: `/?theme=dark#/css/thing
+       *
+       * 2. Adds trailing slash
+       *
+       * Given a route like this: `/#/css/thing`
+       * Will produce a path like this: `/#/css/thing/`
+       */
+      _cleanRawPath: function() {
```

Comment:
Here's where the magic rewriting of the raw path kicks off. Collects all the query params throughout the route and places them at the beginning of the path before the hash (#/foo/) route to match what `app-location` and the native browser query parser expect.